File: tests/DesktopGrass.Win2D.Tests/SimTests/SceneTests.cs
// SceneTests.cs - §13 scene infrastructure.

using DesktopGrass.Win2D;
using Xunit;

namespace DesktopGrass.Win2D.Tests.SimTests;

public class SceneTests
{
    private const double Monitor1920 = 1920.0;

    private static Sim BuildSim(ulong seed = Constants.CANONICAL_TEST_SEED,
                                double monitorWidth = Monitor1920,
                                double density = 1.0)
    {
        var sim = new Sim
        {
            Blades = Sim.GenerateBlades(seed, monitorWidth, density),
            WindowHeight = Constants.STRIP_HEIGHT + Constants.HEADROOM,
            GroundY = Constants.STRIP_HEIGHT + Constants.HEADROOM,
        };
        sim.ResetAmbientGusts(seed, monitorWidth);
        return sim;
    }

    [Fact]
    public void SceneEnumHasSpecLockedDiscriminants()
    {
        Assert.Equal(0, (int)Scene.Grass);
        Assert.Equal(1, (int)Scene.Desert);
        Assert.Equal(2, (int)Scene.Winter);
        Assert.Equal(3, Constants.SCENE_COUNT);
        Assert.Equal(0, (int)Constants.SCENE_DEFAULT);
    }

    [Fact]
    public void FreshSimDefaultsCurrentSceneToGrass()
    {
        var sim = new Sim();
        Assert.Equal(Scene.Grass, sim.CurrentScene);
    }

    [Fact]
    public void SetSceneIsStateOnlyAndDoesNotPerturbBlades()
    {
        var a = BuildSim();
        var b = BuildSim();

        Assert.Equal(a.Blades.Length, b.Blades.Length);
        b.SetScene(Scene.Desert);

        Assert.Equal(Scene.Desert, b.CurrentScene);
        Assert.Equal(Scene.Grass, a.CurrentScene);
        Assert.Equal(a.Blades.Length, b.Blades.Length);
        for (int i = 0; i < a.Blades.Length; i++)
        {
            Assert.Equal(a.Blades[i].BaseX, b.Blades[i].BaseX);
            Assert.Equal(a.Blades[i].Height, b.Blades[i].Height);
            Assert.Equal(a.Blades[i].Thickness, b.Blades[i].Thickness);
            Assert.Equal(a.Blades[i].Hue, b.Blades[i].Hue);
            Assert.Equal(a.Blades[i].IsFlower, b.Blades[i].IsFlower);
            Assert.Equal(a.Blades[i].IsMushroom, b.Blades[i].IsMushroom);
        }
        Assert.Equal(a.AmbientPrng.State, b.AmbientPrng.State);
        Assert.Equal(a.NextAmbientGustTime, b.NextAmbientGustTime);
    }

    [Fact]
    public void SetSceneRoundTripsThroughAllValues()
    {
        var sim = BuildSim(density: Constants.DEFAULT_DENSITY);
        Assert.Equal(Scene.Grass, sim.CurrentScene);
        sim.SetScene(Scene.Desert);
        Assert.Equal(Scene.Desert, sim.CurrentScene);
        sim.SetScene(Scene.Winter);
        Assert.Equal(Scene.Winter, sim.CurrentScene);
        sim.SetScene(Scene.Grass);
        Assert.Equal(Scene.Grass, sim.CurrentScene);
    }

    [Fact]
    public void PerScenePaletteTablesAreFullAlphaArgbEntries()
    {
        for (int s = 0; s < Constants.SCENE_COUNT; s++)
        {
            for (int i = 0; i < Constants.PALETTE_SIZE; i++)
            {
                uint argb = Constants.SCENE_PALETTES[s, i];
                byte alpha = (byte)((argb >> 24) & 0xFFu);
                Assert.Equal((byte)0xFF, alpha);
            }
        }
    }

    [Fact]
    public void GrassScenePaletteIsBitIdenticalToOriginalPalette()
    {
        for (int i = 0; i < Constants.PALETTE_SIZE; i++)
        {
            Assert.Equal(Constants.PALETTE[i], Constants.SCENE_PALETTES[(int)Scene.Grass, i]);
        }
    }

    [Fact]
    public void DesertPaletteValuesMatchSpec()
    {
        uint[] expected =
        {
            0xFFC9A26B, 0xFFB48A56, 0xFFD9B57A,
            0xFF8F6E3F, 0xFFE6C896, 0xFFA67843,
        };

        for (int i = 0; i < Constants.PALETTE_SIZE; i++)
        {
            Assert.Equal(expected[i], Constants.SCENE_PALETTES[(int)Scene.Desert, i]);
            Assert.Equal(expected[i], Constants.DESERT_PALETTE[i]);
        }
    }

    [Fact]
    public void WinterPaletteValuesMatchSpec()
    {
        uint[] expected =
        {
            0xFFE8EEF5, 0xFFB7C4D2, 0xFFCBD8E5,
            0xFFD7E2EE, 0xFFA8B7C6, 0xFFEEF3F8,
        };

        for (int i = 0; i < Constants.PALETTE_SIZE; i++)
        {
            Assert.Equal(expected[i], Constants.SCENE_PALETTES[(int)Scene.Winter, i]);
            Assert.Equal(expected[i], Constants.WINTER_PALETTE[i]);
        }
    }
}

File: tests/DesktopGrass.Native.Tests/src/scene_tests.cpp
// scene_tests.cpp
//
// Scene infrastructure tests (architecture.md §13).
//
// Coverage:
//   * Scene enum discriminants match the spec ({Grass=0, Desert=1, Winter=2, Autumn=3}).
//   * sim_init defaults currentScene to SCENE_DEFAULT (= Grass).
//   * sim_set_scene does not perturb blade positions/dimensions/hues or
//     any non-scene PRNG stream.
//   * Per-scene palette tables are 6 entries each with full-alpha ARGB.
//   * SCENE_PALETTES[Grass] is bit-identical to the original §4 PALETTE.

#include "TestHelpers.h"
#include "Sim.h"
#include "snapshot_data.h"

#include <cstdint>

using namespace desktopgrass;

using namespace Microsoft::VisualStudio::CppUnitTestFramework;

namespace DesktopGrassNativeTests
{
TEST_CLASS(SceneTests)
{
public:
TEST_METHOD(SceneEnumHasSpecLockedDiscriminants) {
    Assert::IsTrue(static_cast<int>(Scene::Grass)  == 0);
    Assert::IsTrue(static_cast<int>(Scene::Desert) == 1);
    Assert::IsTrue(static_cast<int>(Scene::Winter) == 2);
    Assert::IsTrue(static_cast<int>(Scene::Autumn) == 3);
    Assert::IsTrue(static_cast<int>(Scene::Ocean)  == 4);
    Assert::IsTrue(SCENE_COUNT == 5);
    Assert::IsTrue(static_cast<int>(SCENE_DEFAULT) == 0);
}

TEST_METHOD(SimInitDefaultsCurrentSceneToGrass) {
    Sim sim = sim_init(CANONICAL_TEST_SEED, 1920.0, DEFAULT_DENSITY);
    Assert::IsTrue(sim.currentScene == Scene::Grass);
}

TEST_METHOD(SimSetSceneDoesNotPerturbBladeGeometryOrHues) {
    Sim a = sim_init(CANONICAL_TEST_SEED, 1920.0, 1.0);
    Sim b = sim_init(CANONICAL_TEST_SEED, 1920.0, 1.0);

    // Same seed → same blades initially.
    Assert::IsTrue(a.blades.size() == b.blades.size());
    Assert::IsTrue(a.blades.size() == desktopgrass::test::CANONICAL_BLADE_COUNT);

    sim_set_scene(b, Scene::Desert);

    Assert::IsTrue(b.currentScene == Scene::Desert);
    Assert::IsTrue(a.currentScene == Scene::Grass);
    Assert::IsTrue(a.blades.size() == b.blades.size());
    for (size_t i = 0; i < a.blades.size(); ++i) {
        Assert::IsTrue(a.blades[i].baseX     == Near(b.blades[i].baseX));
        Assert::IsTrue(a.blades[i].height    == Near(b.blades[i].height));
        Assert::IsTrue(a.blades[i].thickness == Near(b.blades[i].thickness));
        Assert::IsTrue(a.blades[i].hue       == b.blades[i].hue);
    }
    // Desert cacti may mutate variant tags, but geometry and ambient PRNG stay untouched.
    Assert::IsTrue(a.ambientPrng.state == b.ambientPrng.state);
    Assert::IsTrue(a.nextAmbientGustTime == Near(b.nextAmbientGustTime));
}

TEST_METHOD(SimSetSceneRoundTripsThroughAllValues) {
    Sim sim = sim_init(CANONICAL_TEST_SEED, 1920.0, DEFAULT_DENSITY);
    sim_set_scene(sim, Scene::Desert);  Assert::IsTrue(sim.currentScene == Scene::Desert);
    sim_set_scene(sim, Scene::Winter);  Assert::IsTrue(sim.currentScene == Scene::Winter);
    sim_set_scene(sim, Scene::Autumn);  Assert::IsTrue(sim.currentScene == Scene::Autumn);
    sim_set_scene(sim, Scene::Ocean);   Assert::IsTrue(sim.currentScene == Scene::Ocean);
    sim_set_scene(sim, Scene::Grass);   Assert::IsTrue(sim.currentScene == Scene::Grass);
}

TEST_METHOD(PerScenePaletteTablesAre6ARGBEntriesWithFullAlpha) {
    for (int s = 0; s < SCENE_COUNT; ++s) {
        for (int i = 0; i < PALETTE_SIZE; ++i) {
            const uint32_t argb = SCENE_PALETTES[s][i];
            const uint8_t alpha = static_cast<uint8_t>((argb >> 24) & 0xFFu);
            Assert::IsTrue(alpha == 0xFFu);
        }
    }
}

TEST_METHOD(GrassScenePaletteIsBitIdenticalToTheOriginal4PALETTE) {
    for (int i = 0; i < PALETTE_SIZE; ++i) {
        Assert::IsTrue(SCENE_PALETTES[static_cast<int>(Scene::Grass)][i] == PALETTE[i]);
    }
}

TEST_METHOD(DesertPaletteValuesMatchSpec13) {
    constexpr uint32_t expected[PALETTE_SIZE] = {
        0xFFC9A26Bu, 0xFFB48A56u, 0xFFD9B57Au,
        0xFF8F6E3Fu, 0xFFE6C896u, 0xFFA67843u,
    };
    for (int i = 0; i < PALETTE_SIZE; ++i) {
        Assert::IsTrue(SCENE_PALETTES[static_cast<int>(Scene::Desert)][i] == expected[i]);
        Assert::IsTrue(DESERT_PALETTE[i] == expected[i]);
    }
}

TEST_METHOD(WinterPaletteValuesMatchSpec13) {
    constexpr uint32_t expected[PALETTE_SIZE] = {
        0xFFE8EEF5u, 0xFFB7C4D2u, 0xFFCBD8E5u,
        0xFFD7E2EEu, 0xFFA8B7C6u, 0xFFEEF3F8u,
    };
    for (int i = 0; i < PALETTE_SIZE; ++i) {
        Assert::IsTrue(SCENE_PALETTES[static_cast<int>(Scene::Winter)][i] == expected[i]);
        Assert::IsTrue(WINTER_PALETTE[i] == expected[i]);
    }
}
};
}
